typing/printtyp.ml

```diff
@@ -1581,6 +1578,10 @@ type rec_item_group =
   | Not_rec of syntactic_sig_item
   | Rec_group of (Ident.t list * syntactic_sig_item list)
 
+let with_hidden_items ~is_type ids f =
+  if is_type && not !Clflags.real_paths then
+    wrap_env (hide ids) (Naming_context.with_hidden ids f)
```

Comment:
I am worried about the potential performance impact of `wrap_env`, which is calling `set_printing_env` twice instead of once. I understand that this is key to fixing your bug, but it is also known to be a performance-sensitive aspect of the type-checker. I would like to either:
- convince myself that `wrap_env` is not in fact used much more than before (but I am skeptical here), or
- get a performance test in a short-path-sensitive scenario (I don't know how to do this, but Leo, Fred or Thomas would know), or
- ensure that `wrap_env` does not cost, by caching the short-path-related state across `wrap_env` instead of recomputing it (this should probably be a separate, preliminary PR)

Reply:
I think grouping together the printing environment and short-path states could work well. Then, it would be simpler to cache both at the same time.

Reply:
I added the caching, in a separate commit, I am not completely sure if it necessary to split it from this PR.

Reply:
I would prefer to review this caching as a first separate PR that the printing changes builds on. This is a sensitive change that needs to be reviewed carefully, and it will be yet another occasion to gently ping @lpw25 on the short-paths reimplementation.

Reply:
@Octachron I could be convinced to try to do the separating-the-PRs work next week if you are indeed in holidays during that time. Let me know if that would help.

Reply:
I should have some time  to split the commit by by the end of the week.

Reply:
(gentle ping: you might have a time window open to split the PRs between 4.11.1 and 4.11.2.)

Reply:
This has now be done, the caching logic was reviewed and merged (#9889).